src/shared/local-addresses.c

```diff
@@ -313,3 +316,154 @@ int local_gateways(sd_netlink *context, int ifindex, int af, struct local_addres
+
+                return 0;
+        }
+
+        for (int i = 0; i < n; i++) {
+                _cleanup_close_ int fd = -1;
+                union sockaddr_union sa;
+                socklen_t salen;
+
+                /* Generate exactly one entry per protocol */
+                if (done_af != AF_UNSPEC && af == done_af)
+                        continue;
+
+                fd = socket(a[i].family, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+                if (fd < 0)
+                        return -errno;
+
+                switch (a[i].family) {
+
+                case AF_INET:
+                        sa.in = (struct sockaddr_in) {
+                                .sin_family = AF_INET,
+                                .sin_addr = a[i].address.in,
+                                .sin_port = htobe16(53), /* doesn't really matter wich port we pick — we just care about the routing decision */
+                        };
+
+                        break;
+
+                case AF_INET6:
+                        sa.in6 = (struct sockaddr_in6) {
+                                .sin6_family = AF_INET6,
+                                .sin6_addr = a[i].address.in6,
+                                .sin6_port = htobe16(53),
+                                .sin6_scope_id = a[i].ifindex,
+                        };
+
+                        break;
+
+                default:
+                        assert_not_reached("Unexpected protocol");
+                }
+
+                /* So ideally we'd just use IP_UNICAST_IF here to pass the ifindex info to the kernel before
+                 * connect()ing, sot that it influences the routing decision. However, on current kernels
+                 * IP_UNICAST_IF doesn't actually influence the routing decision for UDP — which I think
+                 * should probably just be considered a bug. Once that bug is fixed this is the best API to
+                 * use, since it is the most lightweight. */
+                r = socket_set_unicast_if(fd, a[i].family, a[i].ifindex);
+                if (r < 0)
+                        log_debug_errno(r, "Failed to set unicast interface index %i, ignoring: %m", a[i].ifindex);
+
+                /* We'll also use SO_BINDTOINDEX. This requires CAP_NET_RAW on old kernels, hence there's a
+                 * good chance this fails. Since 5.7 this restriction was dropped and the first
+                 * SO_BINDTOINDEX on a socket may be done without privileges. This one has the benefit of
+                 * really influencing the routing decision, i.e. this one works for us.*/
+                r = socket_bind_to_ifindex(fd, a[i].ifindex);
+                if (r < 0)
+                        log_debug_errno(r, "Failed to bind socket to interface %i, ignoring: %m", a[i].ifindex);
+
+                if (connect(fd, &sa.sa, SOCKADDR_LEN(sa)) < 0)
+                        log_debug_errno(errno, "Failed to connect SOCK_DGRAM socket to gateway, ignoring: %m");
```

Comment:
is there any point in continuing with `getsockname()` if `connect()` fails?

Reply:
My thinking here was that connect() might ultimately fail for various reasons, for example firewall/policy decisions, there could be anything. But if the routing decision was already made when that failed, then we can still use it. Hence, after calling connect() we check with getsockname() if it is bound now. if so, we use the data regardles if connect() failed. If it's still unbound then we detect that and return a failure.

i.e. i figured it would increase the chance for the whole logic to work, as we can neatly recognize late rwhether the routing decision was made or not.

i'll add a comment about that